duktape/src/main/jni/DuktapeContext.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "DuktapeContext.h"
+#include "JString.h"
+#include "JavaMethod.h"
+
+namespace {
+
+// Internal names used for properties in the duktape context's global stash and bound variables.
+// The \xff\xff part keeps the variable hidden from JavaScript (visible through C API only).
+const char* JAVA_VM_PROP_NAME = "\xff\xffjavaVM";
+const char* JAVA_THIS_PROP_NAME = "\xff\xffjava_this";
+const char* JAVA_METHOD_PROP_NAME = "\xff\xffjava_method";
+
+JNIEnv* getJNIEnv(duk_context *ctx) {
+  duk_push_global_stash(ctx);
+  if (!duk_get_prop_string(ctx, -1, JAVA_VM_PROP_NAME)) {
+    duk_pop(ctx);
+    return nullptr;
+  }
+  JavaVM* javaVM = static_cast<JavaVM*>(duk_require_pointer(ctx, -1));
+  duk_pop_2(ctx);
+
+  JNIEnv* env;
+  javaVM->AttachCurrentThread(&env, nullptr);
+  return env;
+}
+
+jobject getJavaThis(duk_context* ctx) {
+  duk_push_this(ctx);
+  if (!duk_get_prop_string(ctx, -1, JAVA_THIS_PROP_NAME)) {
```

Comment:
Not possible at the moment, no.  Might be a bit over-defensive, I'll undo these checks.